core/src/main/java/com/linecorp/armeria/client/HttpRequestSubscriber.java

```diff
@@ -370,4 +334,68 @@ private boolean cancelTimeout() {
         this.timeoutFuture = null;
         return timeoutFuture.cancel(false);
     }
+
+    private WriteFutureListener writeFutureListener(boolean isRequestHeadersFuture) {
+        if (!isRequestHeadersFuture) {
+            // Reuse in case sending streaming requests.
+            if (cachedWriteFutureListener == null) {
+                cachedWriteFutureListener = new WriteFutureListener(false);
+            }
+            return cachedWriteFutureListener;
+        }
+        return new WriteFutureListener(true);
+    }
+
+    private class WriteFutureListener implements ChannelFutureListener {
+
+        private final boolean isRequestHeadersFuture;
+
+        WriteFutureListener(boolean isRequestHeadersFuture) {
+            this.isRequestHeadersFuture = isRequestHeadersFuture;
+        }
+
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            // If a message has been sent out, cancel the timeout for starting a request.
+            cancelTimeout();
+
+            try (SafeCloseable ignored = RequestContextUtil.pop()) {
+                if (future.isSuccess()) {
+                    // The first write is always the first headers, so log that we finished our first transfer
+                    // over the wire.
+                    if (!loggedRequestFirstBytesTransferred) {
+                        logBuilder.requestFirstBytesTransferred();
+                        loggedRequestFirstBytesTransferred = true;
+                    }
+
+                    if (state == State.DONE) {
+                        logBuilder.endRequest();
+                        // Successfully sent the request; schedule the response timeout.
+                        assert responseWrapper != null;
+                        responseWrapper.initTimeout();
+                    }
+
+                    // Request more messages regardless whether the state is DONE. It makes the producer have
+                    // a chance to produce the last call such as 'onComplete' and 'onError' when there are
+                    // no more messages it can produce.
+                    if (!isSubscriptionCompleted) {
+                        assert subscription != null;
+                        subscription.request(1);
+                    }
+                    return;
+                }
+
+                if (isRequestHeadersFuture) {
```

Comment:
Couldn't we use `loggedRequestFirstBytesTransferred` for `UnprocessedRequestException`?

Reply:
Oops, thanks for finding this.